### ISSUE TITLE
Using --warn-on-type-ref shows MT1502 warnings that can be confusing

### ISSUE THREAD

**Issue by jfversluis** [xamarin/xamarin-macios#8814]
For full backstory see this Forms issue: https://github.com/xamarin/Xamarin.Forms/issues/10967

### Steps to Reproduce

1. Create Forms project on VSWin
2. Do all the flags and right settings to strip out the UIWebView reference (`set --optimize=experimental-xforms-product-type` and `--warn-on-type-ref=UIKit.UIWebView` flags)
3. Build and observe that there are warnings for the type still being found in the binaries

From [various tests](https://github.com/xamarin/Xamarin.Forms/issues/10967#issuecomment-641113073), I have found that the functionality does actually work, but the warnings seem to be false positives.

While talking to @PureWeen we came to the conclusion that the reference probably still shows up on the Windows output folder since they won't contain the linked binaries. While that sounds plausible, one scenario then still seems strange.

A project created on Windows, then copied to a Mac and built with VSMac, still seems to show the reference warning.

### Expected Behavior

Warnings should not be shown as the references have disappeared, or at least some note is shown to the user that the results aren't available/reliable on Windows.

### Actual Behavior

Warnings are shown in the build output while references are actually linked out

### Environment

VSMac
```
Visual Studio Enterprise 2019 for Mac (Preview)
Version 8.7 Preview (8.7 build 1459)
Installation UUID: 685de015-5e4d-4da4-8f89-f84bfe7e8fb1
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 612000074

Mono Framework MDK
Runtime:
	Mono 6.12.0.74 (2020-02/e9d3af508e4) (64-bit)
	Package version: 612000074

Roslyn (Language Service)
3.7.0-3.20273.2+57af2cd3596ec490ee1208ad7285b8f018b1432a

NuGet
Version: 5.6.0.6591

.NET Core SDK
SDK: /usr/local/share/dotnet/sdk/3.1.300/Sdks
SDK Versions:
	3.1.300
	3.1.200
	3.1.102
	3.1.100
	3.1.100-preview3-014645
	3.1.100-preview1-014459
	3.0.100
	3.0.100-rc1-014190
	3.0.100-preview8-013656
	2.2.300
	2.2.203
	2.1.505
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.12.0/lib/mono/msbuild/Current/bin/Sdks

.NET Core Runtime
Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.1.4
	3.1.2
	3.1.0
	3.1.0-preview3.19553.2
	3.1.0-preview1.19506.1
	3.0.3
	3.0.0
	3.0.0-rc1-19456-20
	3.0.0-preview8-28405-07
	2.2.7
	2.2.5
	2.2.4
	2.1.18
	2.1.16
	2.1.15
	2.1.14
	2.1.13
	2.1.9

Xamarin.Profiler
Version: 1.6.13.1
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

Updater
Version: 11

Apple Developer Tools
Xcode 11.5 (16139)
Build 11E608c

Xamarin Designer
Version: 16.7.0.370
Hash: d0db7acb1
Branch: remotes/origin/d16-7
Build date: 2020-05-21 04:20:32 UTC

Xamarin.Mac
Version: 6.20.1.4 (Visual Studio Enterprise)
Hash: 52a30b9fe
Branch: d16-7
Build date: 2020-05-27 14:11:12-0400

Xamarin.iOS
Version: 13.20.1.4 (Visual Studio Enterprise)
Hash: 52a30b9fe
Branch: d16-7
Build date: 2020-05-27 14:11:13-0400

Xamarin.Android
Version: 10.3.99.259 (Visual Studio Enterprise)
Commit: xamarin-android/d16-7/feabf52
Android SDK: /Users/jfversluis/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		5.0 (API level 21)
		7.0 (API level 24)
		7.1 (API level 25)
		8.0 (API level 26)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 29.0.6
SDK Build Tools Version: 29.0.2

Build Information: 
Mono: 075c3f0
Java.Interop: xamarin/java.interop/d16-7@d6024f1
ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
SQLite: xamarin/sqlite/3.31.1@49232bc
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-7@23c4fe0

Microsoft OpenJDK for Mobile
Java SDK: /Users/jfversluis/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

Android SDK Manager
Version: 16.6.0.56
Hash: c4cd6b3
Branch: remotes/origin/dev/joj/167bumpremote~4
Build date: 2020-05-15 22:03:24 UTC

Android Device Manager
Version: 16.6.0.115
Hash: 15b9321
Branch: remotes/origin/dev/joj/bumps~3
Build date: 2020-05-15 22:03:44 UTC

Build Information
Release ID: 807001459
Git revision: f3a5b882b2a011e0e6c3583ddef5091433e6aa5a
Build date: 2020-06-05 13:35:23-04
Build branch: master
Xamarin extensions: f3a5b882b2a011e0e6c3583ddef5091433e6aa5a

Operating System
Mac OS X 10.15.5
Darwin 19.5.0 Darwin Kernel Version 19.5.0
    Thu Apr 30 18:25:59 PDT 2020
    root:xnu-6153.121.1~7/RELEASE_X86_64 x86_64

Enabled user installed extensions
DeepClean 1.2.5
Visual Studio for Mac Extension for Tizen 1.1
Xamarin Hot Reload 20190904.3

```

VSWin

```
Microsoft Visual Studio Enterprise 2019 Int Preview
Version 16.7.0 Preview 3.0 [30125.9.master]
VisualStudio.16.IntPreview/16.7.0-pre.3.0+30125.9.master
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Enterprise

Visual C++ 2019   00435-60000-00000-AA327
Microsoft Visual C++ 2019

ADL Tools Service Provider   1.0
This package contains services used by Data Lake tools

ASA Service Provider   1.0

ASP.NET and Web Tools 2019   16.7.275.39088
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2026502+be7fa794c68ffbe3816d1fa9ba6ed98d83756e39
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.7.275.39088
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.7.275.39088
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.4.6000.0
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   16.7.275.39088
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.4.6000.0
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   3.7.0-3.20273.2+57af2cd3596ec490ee1208ad7285b8f018b1432a
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure HDInsight Azure Node   2.4.6000.0
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.4.6000.0
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   16.0
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.4.6000.0
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.30415.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.79+ge3567815aa.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.7.2 (c19c9c3)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.7.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Snapshot Debugging Extension   1.0
Snapshot Debugging Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62005.01040
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   16.0.20520.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.7.0-3.20273.2+57af2cd3596ec490ee1208ad7285b8f018b1432a
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.10.0.0 for F# 4.7   16.7.0-beta.20230.7+082844d9aaa5d2893680b05a1fe0b77ee18719dd
Microsoft Visual F# Tools 10.10.0.0 for F# 4.7

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions (Preview)   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Kubernetes   1.0
Visual Studio Tools for Kubernetes

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.7.000.200 (master@c631c2f)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.7.0.340 (remotes/origin/d16-7@f61b59c2e)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.6.61 (87c64b8)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   10.3.99.259 (d16-7/feabf52)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 075c3f0
    Java.Interop: xamarin/java.interop/d16-7@d6024f1
    ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
    SQLite: xamarin/sqlite/3.31.1@49232bc
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-7@23c4fe0


Xamarin.iOS and Xamarin.Mac SDK   13.20.0.22 (c2a220d84)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

### Example Project (If Possible)

Project created with VSWin, but gives warnings on VSMac: https://github.com/xamarin/Xamarin.Forms/files/4736423/UIWebKitTest.zip

Project created with VSMac, gives no warnings on VSMac, does give warnings on VSWin: https://github.com/xamarin/Xamarin.Forms/files/4745643/TestShellWebview.zip

**Comment by spouliot** [comment]
> the reference probably still shows up on the Windows output folder since they won't contain the linked binaries.

I'm not 100% sure but I don't think assemblies are copied back to the Windows machine.
Look inside the `.app` for the final versions of the assemblies. This is what will be submitted to Apple (compressed into the `.ipa`)

**Comment by jfversluis** [comment]
Sorry if it wasn't clear :) 

That is what I tried to say. The assemblies are not copied back to Windows, my guess is that is why VS will still show the reference warning.

If I unzip the ipa that is going to Apple, or even send it to Apple, I don't find any reference/do not get any warning email.

**Comment by whitneyschmidt** [comment]
Output from building UIWebKitTest on VSMac, set `--optimize=experimental-xforms-product-type, --warn-on-type-ref=UIKit.UIWebView, -v -v -v -v` : [here](https://gist.githubusercontent.com/whitneyschmidt/72c585fcc37b8f1a1a12309cb6ab707e/raw/513b8aed5925feb7bb328c6e79c363494d705178/UIWebKitTest%2520verbose%2520build%2520output)

Output from building TestShellWebView on VSMac, set `--optimize=experimental-xforms-product-type, --warn-on-type-ref=UIKit.UIWebView, -v -v -v -v` : [here](https://gist.github.com/whitneyschmidt/5b082c23ff81684a6502be78a0ed3726)

I was not able to reproduce building ^ w/o warnings. @jfversluis can you confirm that both flags are set in the project? It looks like they might not have been set (at least in the files you uploaded) since I had to manually add them to the mtouch arguments for TestShellWebview but not for UIWebKitTest.

**Comment by jfversluis** [comment]
Note that you will have to build the Release|iPhone configuration. In this scenario you will get warnings because the optimize flag is overruled since the linker is disabled

**Comment by spouliot** [comment]
> VS will still show the reference warning.

The code that issue the warning is run on the mac (only) so it should not matter. The copying part is only important if you're trying to check *manually* for the symbols on the local (windows) files.

**Comment by stephenhauck** [comment]
This is supposed to be a science and not guesswork.....We build, we add the switches and we still find the references.

I cannot have a process that "guesses" it will be accepted by Apple.

There is no legitimate reason to have it in your assemblies any longer so how about we just take them out of the Xamarin code and then when we see the results of grep or findstr we are SURE there is an issue that needs to be resolved and it WILL BE REJECTED and not "maybe" ?

To reiterate from the thread that spawned this one ....

I start a command prompt
change to the project output directory ..\UIWebKitTest.iOS\bin\iPhone\Release>

run this command "FINDSTR /M /s /i uiwebview . > uiwebviewresults.txt"

get this output in the file
**Xamarin.Forms.Platform.iOS.dll_**

This is an out of the box bare bones Xamarin.Forms Shell app ... there should be no reference whatsoever .....

**Comment by stephenhauck** [comment]
Any update on this, it's a critical problem for us?
Do you need anything more from me ... can I help in any way ?

Thanks!

**Comment by whitneyschmidt** [comment]
@stephenhauck I see that you uploaded the Windows build logs 👍 

Could you also upload the full build logs with verbosity `-v -v -v -v` to the Additional mtouch arguments from your Mac machine (now that we've found a workaround for blocking antivirus problem)?

@spouliot Did we need any other information to investigate?

**Comment by stephenhauck** [comment]
I attached the build output and there was nothing in the logs directory on my Mac, I will check again today.

**Comment by stephenhauck** [comment]
Here are some logs...
[8.0.zip](https://github.com/xamarin/xamarin-macios/files/4831181/8.0.zip)

Still getting the warning and updated VS Mac this morning .. two different projects ... 

**This is a critical issue for us.**

Done building project "UIWebKitTest.iOS.csproj".

Build succeeded.

MTOUCH : warning MT1502: One or more reference(s) to type 'UIKit.UIWebView' already exists inside 'Xamarin.Forms.Platform.iOS, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' before linking
MTOUCH : warning MT1503: One or more reference(s) to type 'UIKit.UIWebView' still exists inside 'Xamarin.Forms.Platform.iOS, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' after linking
    2 Warning(s)
    0 Error(s)

[UIWebKitTest.zip](https://github.com/xamarin/xamarin-macios/files/4831212/UIWebKitTest.zip)

**Comment by stephenhauck** [comment]
Here is my VS Mac information
=== Visual Studio Community 2019 for Mac ===

Version 8.6.5 (build 23)
Installation UUID: 80ff7793-23ec-4ee4-86c6-c39810a3eb76
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 610000104

=== Mono Framework MDK ===

Runtime:
	Mono 6.10.0.104 (2019-12/5d03a6fe116) (64-bit)
	Package version: 610000104

=== Roslyn (Language Service) ===

3.6.0-3.20210.9+4eafdcb1bcbd8d3573f2ba6065e56d9b9ce4f8a3

=== NuGet ===

Version: 5.6.0.6591

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/3.1.301/Sdks
SDK Versions:
	3.1.301
	3.1.300
	3.1.202
	3.1.200
	3.1.102
	3.1.101
	3.1.100
	3.0.101
	3.0.100
	2.1.701
	2.1.700
	2.1.505
	2.1.504
	2.1.503
	2.1.302
	2.1.301
	2.1.4
	2.0.0
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.10.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.1.5
	3.1.4
	3.1.2
	3.1.1
	3.1.0
	3.0.1
	3.0.0
	2.1.19
	2.1.18
	2.1.17
	2.1.16
	2.1.15
	2.1.14
	2.1.13
	2.1.12
	2.1.11
	2.1.9
	2.1.8
	2.1.7
	2.1.2
	2.1.1
	2.0.5
	2.0.0

=== Xamarin.Profiler ===

Version: 1.6.15.68
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Xamarin Designer ===

Version: 16.6.0.329
Hash: d4f8bcd13
Branch: remotes/origin/d16-6
Build date: 2020-04-24 02:16:02 UTC

=== Apple Developer Tools ===

Xcode 11.5 (16139)
Build 11E608c

=== Xamarin.Mac ===

Version: 6.18.2.1 (Visual Studio Community)
Hash: 29c4ea731
Branch: d16-6
Build date: 2020-05-26 17:03:04-0400

=== Xamarin.Android ===

Version: 10.3.1.4 (Visual Studio Community)
Commit: xamarin-android/d16-6/3a10de9
Android SDK: /Users/stephenhauck/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		5.1 (API level 22)
		6.0 (API level 23)
		7.0 (API level 24)
		7.1 (API level 25)
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 29.0.6
SDK Build Tools Version: 29.0.2

Build Information: 
Mono: 165f4b0
Java.Interop: xamarin/java.interop/d16-6@2cab35c
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.31.1@49232bc
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-6@bfb66f3

=== Microsoft OpenJDK for Mobile ===

Java SDK: /Users/stephenhauck/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 16.6.0.50
Hash: 5901879
Branch: remotes/origin/d16-6
Build date: 2020-06-10 22:42:50 UTC

=== Android Device Manager ===

Version: 16.6.0.96
Hash: 6e8b80b
Branch: remotes/origin/d16-6
Build date: 2020-06-10 22:43:28 UTC

=== Xamarin.iOS ===

Version: 13.18.2.1 (Visual Studio Community)
Hash: 29c4ea731
Branch: d16-6
Build date: 2020-05-26 17:03:05-0400

=== Build Information ===

Release ID: 806050023
Git revision: 5289d413b99fddfc20b4ecf3e445ccb822213427
Build date: 2020-06-18 12:08:30-04
Build branch: release-8.6
Xamarin extensions: 5289d413b99fddfc20b4ecf3e445ccb822213427

=== Operating System ===

Mac OS X 10.15.5
Darwin 19.5.0 Darwin Kernel Version 19.5.0
    Tue May 26 20:41:44 PDT 2020
    root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64

=== Enabled user installed extensions ===

MFractor 4.2.6

**Comment by stephenhauck** [comment]
Here is my Xcode info
![image](https://user-images.githubusercontent.com/12648793/85721719-9ef3ad80-b6bf-11ea-97d0-c901a451303d.png)

**Comment by stephenhauck** [comment]
Here is my OSX Info
![image](https://user-images.githubusercontent.com/12648793/85721814-b29f1400-b6bf-11ea-8b8a-b26536f9c98d.png)

**Comment by chamons** [comment]
The version info is useful, but I don't see a full build log (full verbosity) and/or an example project.

**Comment by stephenhauck** [comment]
The logs were attached that were generated when the extra parameters were added to the project.
I have previously (in this thread or the one it spawned from) described the steps to reproduce the issue and the source code.
I am sorry but someone with familiarity with the generation and structure of projects in VS Windows and VS Mac plus some MS Build knowledge could probably locate the difference / issue in a matter of hours.

I don;t mean to be rude or curt but this is a critical issue for anyone that creates a project on a Windows machine using Visual Studio and is easily reproducible, close it if you want to.

**Comment by stephenhauck** [comment]
Anyone ....

**Comment by stephenhauck** [comment]
Can I pay an MS engineer to spend 30 minutes resolving this issue ?

**Comment by chamons** [comment]
This issue is on our triage queue still. Between the work of Xcode 12 and recent traffic, I'm sorry it has not been resolved as of yet.

As noted on [our forums](https://forums.xamarin.com/discussion/150918/the-microsoft-xamarin-support-team-is-here-to-help#latest):

> If you want one on one support via email (business hours only) or need to provide private code or other sensitive info, then please submit a support request to the Xamarin support team at [this link](https://support.microsoft.com/en-us/supportforbusiness/productselection?sapId=211dd84f-3474-c3c5-79bf-66db630c92a6).

**Comment by stephenhauck** [comment]
I appreciate it's in "triage" that but it's easily reproduced ...
Is there anything else I can do to speed an answer to this ?

**Comment by stephenhauck** [comment]
I REALLY need a resolution to this issue ....

I updated everything ....
Created a new Shell app for Android and iOS .. NO NEW CODE ADDED!
Made the changes to the iOS build like so 
<img width="967" alt="Screen Shot 2020-07-13 at 7 28 56 AM" src="https://user-images.githubusercontent.com/12648793/87299727-8f0b0480-c4da-11ea-9f7e-30f957ca3fa1.png">

Right clicked and selected "Rebuild.." like so 
<img width="661" alt="Screen Shot 2020-07-13 at 7 33 40 AM" src="https://user-images.githubusercontent.com/12648793/87300068-32f4b000-c4db-11ea-819e-06e0df58520d.png">


Still get the warnings ...
_Done building project "StandardShellApp_CreatedOnMac_7_13_20.iOS.csproj".

Build succeeded.

MTOUCH : warning MT1502: One or more reference(s) to type 'UIKit.UIWebView' already exists inside 'Xamarin.Forms.Platform.iOS, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' before linking
MTOUCH : warning MT1503: One or more reference(s) to type 'UIKit.UIWebView' still exists inside 'Xamarin.Forms.Platform.iOS, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' after linking
    2 Warning(s)
    0 Error(s)_

[LogsFromMac.zip](https://github.com/xamarin/xamarin-macios/files/4912113/LogsFromMac.zip)



[Mac_iOS_Build.zip](https://github.com/xamarin/xamarin-macios/files/4912115/Mac_iOS_Build.zip)

[vsmacdetails.zip](https://github.com/xamarin/xamarin-macios/files/4912116/vsmacdetails.zip)
[StandardShellApp_CreatedOnMac_7_13_20.zip](https://github.com/xamarin/xamarin-macios/files/4912120/StandardShellApp_CreatedOnMac_7_13_20.zip)

**Comment by stephenhauck** [comment]
OK, ran an extra test ...

This works ... (on VS Mac)
![image](https://user-images.githubusercontent.com/12648793/87301181-1bb6c200-c4dd-11ea-9168-8a74def2dd60.png)

This does not .... (on VS Mac)
![image](https://user-images.githubusercontent.com/12648793/87301154-0e99d300-c4dd-11ea-98a2-1b300cef5c0d.png)

**Comment by chamons** [comment]
So I did some poking around on this, and this is what I found:

- I can reproduce warn-on-type-ref warning on Debug by default, which makes sense as by default Debug doesn't enable the linker (and thus we don't drop the type)
- The trouble is that `--optimize=experimental-xforms-product-type` is the pre-released version of the flag. In the [release notes](https://docs.microsoft.com/en-us/xamarin/ios/release-notes/13/13.16) it is noted as now `--optimize=force-rejected-types-removal` which also works for me in both Debug and Release.

Give that a try and report back.

**Comment by tschlarman** [comment]
Sorry for chiming in, but I have the same issue and have been digging around on this for the past couple of days and just found this thread (and the one that spawned it). My project started on and is still on Windows. It also uses the new XF Shell (if it matters). I added in the linker flag that was mentioned.

![image](https://user-images.githubusercontent.com/10763706/88860244-25fcdf80-d1b0-11ea-8b38-9ec5240142ba.png)

After a clean/rebuild:
![image](https://user-images.githubusercontent.com/10763706/88860284-3a40dc80-d1b0-11ea-87b3-b20d88e3596b.png)

In my project folder:
findstr /s /i /m UIWebView *.* 
![image](https://user-images.githubusercontent.com/10763706/88860451-9c99dd00-d1b0-11ea-93c5-c801281b09f2.png)

This should take care of finding anything in the 3rd party iOS specific libs I use. It didn't appear to.

Various environment info:
![image](https://user-images.githubusercontent.com/10763706/88860689-2b0e5e80-d1b1-11ea-81a9-7e077c1cf8f8.png)

iOS project nuget info:
![image](https://user-images.githubusercontent.com/10763706/88860766-55601c00-d1b1-11ea-9618-bbb7929ccaca.png)

I use VS to push to the AppStore right now. When I choose the Archive option I don't see any warnings that indicate a UIWebView issue.

Distributing/uploading to the AppStore seems fine.

Until a few minutes later I get the Dear Developer email...
![image](https://user-images.githubusercontent.com/10763706/88861512-fa2f2900-d1b2-11ea-89b1-2e18dec4543c.png)

Confirming their scan, I Open the folder and copy the *.ipa file to a new folder and unzip it. Again run the findstr command in that folder and it lists: Payload\<myproject>.iOS.app\<myproject>.iOS

I used VS Find All in my solution, so it's not me or any code I have access to.
![image](https://user-images.githubusercontent.com/10763706/88862048-2ac39280-d1b4-11ea-9850-c34905ebd9d8.png)

I'm not sure if any of this helps, other than to verify it's not specific to a machine, it's reproducible on a machine that is running all the current software with the current flags from the release notes and it's not just one person seeing it. I haven't done as much digging @stephenhauck who has been very thorough, but we've both come back to the same spot I think.

I did try building with the -v -v -v -v mtouch option but got some sort of a compiler error saying it didn't know the -v option.

If I can help further let let me know as this is a release blocker at the moment since we can't side-load a file to our testers very easily.

Thanks for any help anyone can provide.

**Comment by stephenhauck** [comment]
I did reach out to the people I know having the issue and they have to build and release on a Mac to get it to pass ...
It's time to put someone on this issue and get it resolved!!!!

**Comment by stephenhauck** [comment]
Here are the latest tests .....
[WinCreated_7_29_20.zip](https://github.com/xamarin/xamarin-macios/files/4998107/WinCreated_7_29_20.zip)
[StandardShellApp_CreatedOnMac_7_13_20.zip](https://github.com/xamarin/xamarin-macios/files/4998108/StandardShellApp_CreatedOnMac_7_13_20.zip)

**Comment by tschlarman** [comment]
Hi @stephenhauck 

I pulled down your 2 files. As I'm sure you know there are only a couple of differences in the csproj files between the 2 projects.

It looks like in Release|iPhone there are no changes between the 2. Only in Release|iPhone Simulator where the mac csproj has the updated params and linker setting and the windows csproj doesn't.

No other changes of note in that file or any other for that matter.

The leads us to what VS generates for msbuild is somehow different between VS mac and VS 2019. I'll see if I can get a capture of that maybe tomorrow. We'll see where that leads us.

**Comment by stephenhauck** [comment]
@tschlarman, yep those are the only differences besides one project was created on a Mac and one on Windows 10.

The items you listed are the same items I have felt from the beginning were the issue but just like me .... if the QA people or the business / user can't reproduce the issue then it doesn't exist.

While I feel your pain and feel for you I am happy that someone else is experiencing this issue.

**Comment by chamons** [comment]
Ok, I did some digging and I believe this might be expected.

Let's compare MT1502 and MT1503 in our docs:

**Comment by chamons** [comment]
Now I will agree the UX of this is less than idea. I'll let @spouliot comment when he returns, so I won't close this.

Now you can force MT1502 errors to be ignore via `--nowarn:1502` and optionally add `--warnaserror:1503` to convert those to errors.

**Comment by stephenhauck** [comment]
if you look at @tschlarman post above he still got the warning from Apple on submission.

Were not crazy .. something's wrong ...

**Comment by tschlarman** [comment]
@chamons I don't think the essence isn't whether the messages are confusing. The essence of the issue is that we can't publish a Xamarin app to the Apple AppStore from VS2019 on windows and get it approved because of the UIWebView reference inside Xamarin Forms. 

We have some circumstantial evidence that they can be built and published under mac and get approved. I'm trying that out today.

**Comment by stephenhauck** [comment]
If I choose "release" build then why isn't the extra "mtouch" command line option simply applied by Microsoft by default ?

**Comment by chamons** [comment]
@tschlarman - Can you post a sample project showing this? 

I tried @stephenhauck 's sample projects and didn't see the warning post link. Confirm that you are using the newer optimziation (non prototype) flag.

You need `force-rejected-types-removal`.

**Comment by stephenhauck** [comment]
@chamons, if you scroll back up you can see he has that in his additional mtouch arguments ... 
![image](https://user-images.githubusercontent.com/12648793/88966466-5b6dff80-d27a-11ea-983a-31b4cbcd44e6.png)

**Comment by chamons** [comment]
I understand that he has it in his args, that doesn't help me reproduce his issue and look into it. I can't determine if the issue is user error or a bug, and fix that bug if it exists, without being able to see it.

**Comment by stephenhauck** [comment]
@chamons, it's not about the warning post link .. it's about the fact that the reference is still in the app and Apple rejects it .
![image](https://user-images.githubusercontent.com/12648793/88966860-fbc42400-d27a-11ea-8cf2-e6244d91c0da.png)

**Comment by stephenhauck** [comment]
@chamons, can you please change the title accordingly ...

Build on VS Mac and Apple accepts it
Build same app on VS Windows and Apple rejects it because the deprecated feature remains in the ipa file somehow

**Comment by chamons** [comment]
Here's what I'm trying to get across:

MT1502 = Noise assuming you have force-rejected-types-removal (it warns on usage before linking)
MT1503 = If reports you will likely be rejected.

The important thing is not being rejected. The 1503 warning points towards that, but is not necessarily that. 

The samples attached here only show 1502 and to my knowledge are not rejected.

If someone can attach a sample that uses force-rejected-types-removal and either shows 1503 AND or is rejected, I can dig into what's going on. 

There should be no difference between VS Mac and VS Windows on this. If someone tells me "this sample works, but only on macOS not Windows" I can work with that.

**Comment by stephenhauck** [comment]
@chamons, thank you for taking time to look into this ...

This issue has been going on for far too long already ...
It just too hard ... I will simply keep putting my apps onto the Mac and building for distribution there ...
Eventually it will be fixed somewhere down the road ....

You can close the issue as far as I'm concerned.

**Comment by tschlarman** [comment]
So building my same project on macos, VSmac says the publish succeeded, but I get the same Dear Developer email complaining about UIWebView. So just a simple rebuild on macos doesn't appear to work in my case.

**Comment by tschlarman** [comment]
Back on windows. If this is not the command line you'd like me to use, please reply with what I should be using verbatim.

![image](https://user-images.githubusercontent.com/10763706/88976307-a97e0580-d270-11ea-99fa-0895824fdd71.png)

and the warning:
![image](https://user-images.githubusercontent.com/10763706/88976925-cbc45300-d271-11ea-8222-3a016fc4aedd.png)

It is not listed as an error so does that mean UIWebView is a pre-linker warning and is not supposed to be in the output? 

If that's the case, findstr says otherwise:
![image](https://user-images.githubusercontent.com/10763706/88977154-337a9e00-d272-11ea-96e4-48b0da64f1b2.png)

**Comment by chamons** [comment]
@tschlarman - :thinking: - warnaserror should for a build error if the type remains. If it doesn't, i don't know what Apple is finding. 

Given the fact that thousands of forms developers aren't pounding my door down, and I haven't been able to reproduce it yet, there must me some condition I'm not fulfilling yet. 

I'm going to try a few more things, but a sample project is worth a thousand words here.

**Comment by tschlarman** [comment]
Just a thought. Since 'Link All' is specified, should the file Xamarin.Forms.Platform,iOS.dll exist at all?  Shouldn't it be linked in as a part of the main app?

**Comment by tschlarman** [comment]
I created a new app in the app store, used the stock Xamarin shell template and it uploaded from VS2019 on Windows just fine.

Now I'll start by adding in the dependencies and see what happens.

**Comment by tschlarman** [comment]
@chamons Thanks for your help. I fell back to creating a test app and added all the dependencies 1 by 1 and kept submitting to the app store until I found the one that caused the issue. Sent the issue to the third party for fixing. I wonder why their libraries didn't register the UIWebView usage.

**Comment by chamons** [comment]
Anyone else have this issue and want to spend some time tracking it down, or should I close this issue?

**Comment by chamons** [comment]
I did submit a Forms app to try to reproduce locally and didn't have any luck, so right now I think it's project specific somehow.

**Comment by stephenhauck** [comment]
So .... 

Because the additional settings string needs to be added to the project ANYONE can build an assembly and share it and create this problem for anyone using Xamarin and building for iOS ?

How / what setting do we put into the additional mtouch arguments that strips any referenced assemblies as well to guarantee it is accepted by Apple ?

**Comment by stephenhauck** [comment]
@ chamons, @ tschlarman .. how do we "scan" NuGet packages and such ?

Trial and error is not effective or repeatable.

**Comment by tschlarman** [comment]
Even though mine appears to be a 3rd party thing, maybe a new issue could be opened for the VS XF templates that would include the above mtouch arguments and linker setting by default. That would help with any new folks having to go thru some of the same ground we went through just to end up here. 

Another fix that could be helpful is an update for the top ranked blog entry on this topic by @jfversluis that still shows the old --optimize entry. It needs to be updated to show the new flag from the release notes. (https://devblogs.microsoft.com/xamarin/uiwebview-deprecation-xamarin-forms/)

This would at least save people some research and trial and error time.

Still the best advice (but not time effective or any fun) on this issue seems to be: create a new project, add the project settings above and submit to Apple. If it passes (it should), add all your dependencies 1 by 1 and submit to Apple after each until something fails. Then you know who to contact.

**Comment by stephenhauck** [comment]
Why doesn't the linker remove deprecated content automatically ... the additional settings and the ability to create multiple build configurations it's a management issue that can be resolved by Microsoft simply adding it as a setting in Visual Studio project setting

**Comment by tschlarman** [comment]
Sounds like a really good idea. The pit of success.

**Comment by chamons** [comment]
Let me attempt to clarify for you @stephenhauck .

- At one point Apple provided UIWebView for programmers to use
- They later decided it was not good anymore, and reject applications that use it.
- Xamarin.iOS projects tend to be linker safe, so unless your application uses the type directly, linker enabled builds will strip it away. Xamarin.Forms however is not linker safe, at least currently. 
- Libraries that have strict API compatibility, such as Xamarin.iOS and Xamarin.Forms can not remove the type. 
- This means C# libraries and applications can reference UIWebView, which will be rejected at submission time. This specifically happens most often in Forms applications.

`force-rejected-types-removal` and `warn-on-type-ref` are our (the SDK) two part solution to address Apple's hard removal of a commonly used type.

- [force-rejected-types-removal](https://github.com/xamarin/xamarin-macios/pull/8009) tells the linker to rename the type in your application, which will cause all references to UIWebView to both:
    - Pass the Apple check
    - Fail at runtime if used
- warn-on-type-ref will yell if an assembly references has a reference before or after linking, so you can track down which _managed_ assemblies reference the now Apple forbidden type.

If you have 3rd party libraries, specifically native one, it is possible that force-rejected-types-removal is insufficient. Any reference to the type, directly in your code, in a library that you depend on, etc, will cause Apple to reject your application.

No one here is able to reproduce the issue to know if either:
- There is a corner case in force-rejected-types-removal that it is missing a managed reference
OR
- There is something project specific in your application that is triggering this.

As of now, there is no evidence that there is "trial or error" or that macOS vs Windows has any effect. 

If you'd like to suggest Xamarin.Forms change their template feel free to [file an issue](https://github.com/xamarin/Xamarin.Forms/issues/new).

I'll suggest updating the blog post, that's a good idea.

If you want assistance, we need a sample project showing your specific issue to proceed.

**Comment by stephenhauck** [comment]
Thanks for the detailed reply !

"trial and error" refers to adding and submitting until failure is reached ... This is impossible on an existing code base.

Is there not some tool (even if it costs money) to look at all references even the native libraries ?

you mentioned .."Libraries that have strict API compatibility, such as Xamarin.iOS and Xamarin.Forms can not remove the type."

Why not ?

If Apple has removed it then you remove it and create a hard error ... that's the stopping point for people using deprecated API's.

If I have time to create an example that fails I will look this thread back up but right now the hack seems to be to build on a Mac for some reason.

**Comment by tschlarman** [comment]
@chamons thank for the insight. The native reference explains why my 3rd party issue got by and the linker warning didn't scream at it. Agreed this doesn't seem to be a Mac vs Windows thing nor is it a project thing AFAICT.

Thanks for the link file an issue. I will do that.

@chamons thanks for being patient with us and we try and slog thru this.

**Comment by chamons** [comment]
Reasonable questions.

> Libraries that have strict API compatibility, such as Xamarin.iOS and Xamarin.Forms can not remove the type. - Why?

If you change the contract for a .NET assembly, by removing existing methods or types, entire libraries may no longer be loadable. Some of those libraries may be commercial libraries, or the source unavailable for rebuild, or nugets with many downloads. Customers many have no option to move forward until their dependencies also move forward. 

Thus forcing an API break is flatly unacceptable for our customers, unless Apple strictly forces it (such as the 32-bit to 64-bit migration a number of years back). 

> Why doesn't the linker remove deprecated content automatically

Unless something is a hard build break, where applications can not build or link (example: QTKit removal on macOS), we tend to follow Apple's lead here. 

There could be enterprise customer shipping applications via channels other than the App Store using that type, and we'd be more strict than Apple here. 

Now if Apple actually removed the types from binaries, or dropped the headers, we'd reconsider.

**Comment by tschlarman** [comment]
> Customers many have no option to move forward until their dependencies also move forward.
This is pretty typical in the Apple ecosystem however. Look how many times they've broken Swift compatibility.

**Comment by chamons** [comment]
Yep - but when Xamarin can do better we absolutely strive to.

This get complicated with enterprise accounts, and there is more than one way of shipping internal apps. I believe newer / preferred was is via the app store, but I know there are other/older options.

**Comment by tschlarman** [comment]
Agreed. Sorry for getting off topic.  😃 

Thank you @chamons for your insight. Much appreciated.

**Comment by PureWeen** [comment]
@chamons 

>  Xamarin.Forms however is not linker safe, at least currently.

I don't think this is true
https://github.com/xamarin/Xamarin.Forms/blob/main/Xamarin.Forms.Platform.iOS/Properties/AssemblyInfo.cs#L65
It's been Linker Safe since 4.5

> force-rejected-types-removal 

AFAIK this flag won't have any effect on Xamarin.Forms itself. The only reason to add this flag would be if you are using a 3rd party library that still uses UIWebview  but in that case you might get a runtime exception. So does it widen the pit of success? Or does it just shift the same pit to the left a little bit?

> warn-on-type-ref

I could see possibly adding this flag though there are a lot of comments here about that flag and I'm not sure if just adding it will have the intended result? Was @jfversluis initial concern addressed? If we add this flag will it tell users what 3rd party library is referencing UIWebView?

> Existing templates

There's a few things going on here that are mainly just due to a lot of pre-release bits that haven't caught up with stable yet

1) the Xamarin.Forms templates are using a version of XF 4.5 that isn't linker safe but if you update to the latest XF 4.5 or 4.7 it is linker safe. VS 8.7 will be pointing at XF 4.7
2) The following flag is all you really need to set as it relates to XF `--optimize=experimental-xforms-product-type`. This flag is on by default in VS 8.7 right @chamons ? 

So once VS 8.7 stable lands then everything should just link out using the default templates and changing nothing

 I tested the following steps on 8.6 
- creating a Xamarin forms shell app on vsmac
- changed the settings to Release/Generic Device
- updated to the latest XF 4.5
- set this flag `--optimize=experimental-xforms-product-type`
- right click "Archive for publishing"

If you check the dlls produced by this process they don't include a UIWebView

So I think most of this will all just be resolved once VS 8.7 hits stable

**Comment by chamons** [comment]
That is great news @PureWeen! I was out of the loop on the forms specific bits. 

Sebastien handled most of that, and he's out right now so I couldn't double check before my posts.

**Comment by PureWeen** [comment]
I tested with the recently stable VS MAC 8.7 and the`--optimize=experimental-xforms-product-type` flag was no longer needed.

The default XF template with VS MAC 8.7 should now strip UIWebView out with no additional user steps required.

**Comment by stephenhauck** [comment]
OK so I updated VS on the Mac and VS on Windows 10
I took an example app from Microsoft that accesses Azure (attached)
